http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpRequest.java

```diff
@@ -304,7 +304,8 @@ protected Object buildBody() {
             for (ByteBufHolder holder : receivedContent) {
                 ByteBuf content = holder.content();
                 if (content != null) {
-                    byteBufs.addComponent(true, content);
+                    // need to retain content, because for addComponent "ownership of buffer is transferred to this CompositeByteBuf."
+                    byteBufs.addComponent(true, content.retain());
```

Comment:
The content is already retained when the `addContent` method is called. Can you explain why it needs to be retained again and how retaining the content twice wouldn't lead to a memory leak without an additional release?

Reply:
The retain in addContent is on the attribute. It does not affect the buffer itself.

The release of the attribute in receivedContent is done in NettyHttpRequest.release

Reply:
https://github.com/micronaut-projects/micronaut-core/blob/3.3.x/http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java#L891